### PR TITLE
Refactor HCP Vault TF acceptance tests for Azure

### PR DESCRIPTION
**Pull request by codergs**
This PR adds acceptance tests with Azure cloud provider. Not all features are available under Azure cloud.

refactor is done so that we can use same code base for AWS And Azure cloud
separated out perf replication tests to a separate file - no refactor is done for this at the moment.
fixed a regression on perf replication test suites

### :building_construction: Acceptance tests

- [ ] Are there any feature flags that are required to use this functionality?
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
Wed Oct 12 16:48:43 EDT 2022 hashicorp/terraform-provider-hcp-internal - (codergs-HVT-2273) > TF_LOG_PATH=out.text TF_LOG=TRACE TF_ACC=1 go test ./internal/... -parallel 20 -v -run=TestAccVaultClusterAWS -timeout 210m     
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/clients    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/consul     (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/input      (cached) [no tests to run]
=== RUN   TestAccVaultClusterAWS
=== PAUSE TestAccVaultClusterAWS
=== CONT  TestAccVaultClusterAWS
--- PASS: TestAccVaultClusterAWS (2630.25s)
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/provider   2630.911s
```

### CODE REVIEW

File: internal/provider/resource_vault_cluster_perf_replication_test.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+				resource "hcp_vault_cluster" "c2" {
+					cluster_id   = "test-secondary"
+					hvn_id       = hcp_hvn.hvn1.hvn_id
+					tier         = hcp_vault_cluster.c1.tier
+					primary_link = hcp_vault_cluster.c1.self_link
+					paths_filter = ["path/a", "path/c"]
+				}
+				`)),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(secondaryVaultResourceName, "paths_filter.0", "path/a"),
```
Comment (bcmdarroch) [inline internal/provider/resource_vault_cluster_perf_replication_test.go:218]: Test design question: I've been worried about the amount of time this test takes-- typically over an hour. I've often run into timeouts when testing the entire acceptance suite locally.

I appreciate that you've added the Vault cluster tests in parallel. :) But looking at this performance replication test, I wonder if we might save some test time by simplifying the steps to cover a single happy path, like a single field update. The other edge cases should certainly be covered, but with the provider being responsible for testing so many different HCP resources on every release, this repo might not be the best place for them.

File: internal/provider/resource_vault_cluster_test.go [hunk 2]
```diff
@@ -287,325 +135,183 @@ func testAccCheckVaultClusterDestroy(s *terraform.State) error {
...
 }
 
-func setTestAccPerformanceReplicationE2E(vaultCluster string) string {
-	return fmt.Sprintf(`
-resource "hcp_hvn" "hvn1" {
-	hvn_id            = "test-perf-hvn-1"
-	cidr_block        = "172.25.16.0/20"
-	cloud_provider    = "aws"
-	region            = "us-west-2"
+// utlity functions
```
Comment (bcmdarroch) [inline internal/provider/resource_vault_cluster_test.go:138]:
```suggestion
// utility functions
```

File: internal/provider/resource_vault_cluster_perf_replication_test.go (hunk 1 above)
Comment (bcmdarroch) [reply internal/provider/resource_vault_cluster_perf_replication_test.go:218]: Our team is working on ways we might cache acceptance test results, so that the entire suite need not be run for unrelated changes. That may help alleviate this issue as well.

Comment (codergs) [reply internal/provider/resource_vault_cluster_perf_replication_test.go:218]: @bcmdarroch Lets sync tomorrow on this. I hear your concern, but I know we want to test all of these paths. Now can choose to do it for one provider, but our AWS and Azure code paths are not same hence we need to test both. Maybe we can find common ground to change some of our test suites to make the experience better. In PRDE it was really hard to test for me as well. not sure if you tested in any other env and had this experience.

Comment (codergs) [reply internal/provider/resource_vault_cluster_perf_replication_test.go:218]: After chatting with @bcmdarroch and @himran92, we decided to keep moving forward. @bcmdarroch and team will look into running per replication test optionally vs every time so that pipeline is not blocked.

File: internal/provider/resource_vault_cluster_const_test.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	cluster_id         = "test-vault-cluster-{{ .CloudProvider }}"
+	hvn_id             = hcp_hvn.test.hvn_id
+	tier               = "{{ .Tier }}"
+}
+`
+
+// sets public_endpoint to true and add metrics and audit log
+const updatedVaultClusterPublicAndMetricsAuditLog = `
+resource "hcp_vault_cluster" "test" {
+	cluster_id         = "test-vault-cluster-{{ .CloudProvider }}"
```
[the remark this answers is not in the document]
Comment (codergs) [reply internal/provider/resource_vault_cluster_const_test.go:33]: @aoripov possible to float a quick update to the format? Should be an easy turnaround. Leaving to you as you are working on the merge of this. If not, then I can get to it by my EOD.

[the remark this answers is not in the document]
Comment (aoripov) [reply internal/provider/resource_vault_cluster_const_test.go]: Added timestamp to Cluster and HVN ids.

**Comment by aoripov** [comment]
@bcmdarroch I ran the tests locally:
```
~/projects/terraform-provider-hcp (codergs-HVT-2273)$ TF_LOG_PATH=out.text TF_LOG=TRACE TF_ACC=1 go test ./internal/... -parallel 20 -v -run=TestAccVaultClusterAWS -timeout 210m
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/clients	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]
=== RUN   TestAccVaultClusterAWS
=== PAUSE TestAccVaultClusterAWS
=== CONT  TestAccVaultClusterAWS
--- PASS: TestAccVaultClusterAWS (4451.42s)
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider	4452.169s
```

File: internal/provider/resource_vault_cluster_test.go [hunk 4]
```diff
@@ -3,229 +3,78 @@ package provider
...
-	tier               = "standard_small"
-	major_version_upgrade_config {
-		upgrade_type = "SCHEDULED"
-		maintenance_window_day = "WEDNESDAY"
-		maintenance_window_time = "WINDOW_12AM_4AM"
+// This includes tests against both the resource, the corresponding datasource, and the dependent admin token resource
+// to shorten testing time.
+func TestAccVaultClusterAzure(t *testing.T) {
+	azureTestInput := inputT{
+		vaultClusterName:           addTimestampSuffix("test-vault-cluster-azure-"),
```
Comment (bcmdarroch) [inline internal/provider/resource_vault_cluster_test.go:33]: Alas this is just over the 36 character limit 😩 
```suggestion
		vaultClusterName:           addTimestampSuffix("test-vault-azure-"),
```

File: internal/provider/resource_vault_cluster_perf_replication_test.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+	hvn1ResourceName           = "hcp_hvn.hvn1"
+	hvn2ResourceName           = "hcp_hvn.hvn2"
+	primaryVaultResourceName   = "hcp_vault_cluster.c1"
+	secondaryVaultResourceName = "hcp_vault_cluster.c2"
+)
+
+func setTestAccPerformanceReplicationE2E(tfCode string) string {
+	return fmt.Sprintf(`
+resource "hcp_hvn" "hvn1" {
+	hvn_id            = "test-perf-hvn-1"
```
Comment (bcmdarroch) [inline internal/provider/resource_vault_cluster_perf_replication_test.go:21]: These IDs also need to be timestamped, with max 36 characters.

Comment (aoripov) [reply internal/provider/resource_vault_cluster_perf_replication_test.go]: Done

File: internal/provider/resource_vault_cluster_test.go (hunk 4 above)
Comment (aoripov) [reply internal/provider/resource_vault_cluster_test.go]: Done

**Comment by aoripov** [comment]
Azure tests:
```
TF_LOG_PATH=out.main.text TF_LOG=TRACE TF_ACC=1 go test ./internal/... -parallel 20 -v -run=TestAccVaultClusterAzure -timeout 240m
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/clients	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]
=== RUN   TestAccVaultClusterAzure
=== PAUSE TestAccVaultClusterAzure
=== CONT  TestAccVaultClusterAzure
--- PASS: TestAccVaultClusterAzure (4493.64s)
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider	4494.746s
```

**Comment by aoripov** [comment]
AWS tests.
Ignore failure. It failed on the last step of deleting an HVN because of an increased cadence latency.
```
TF_LOG_PATH=out.main.text TF_LOG=TRACE TF_ACC=1 go test ./internal/... -parallel 20 -v -run=TestAccVaultClusterAWS -timeout 240m
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/clients	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]
=== RUN   TestAccVaultClusterAWS
=== PAUSE TestAccVaultClusterAWS
=== CONT  TestAccVaultClusterAWS
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1

        Error: unable to delete HVN (test-hvn-aws-202301251324): context canceled waiting for delete HVN operation (5a81e5f1-b4b8-4a46-b801-8ef0fdb42c8b) to complete

--- FAIL: TestAccVaultClusterAWS (7092.17s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-hcp/internal/provider	7092.827s
FAIL
```

**Comment by aoripov** [comment]
Performance replication tests. Same as with AWS tests they failed on the very last step - HVN deletion.
```
TF_LOG_PATH=repl.1.text TF_LOG=TRACE TF_ACC=1 go test ./internal/... -parallel 20 -v -run=TestAccPerformanceReplication_ValidationsAws -timeout 240m
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/clients	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]
=== RUN   TestAccPerformanceReplication_ValidationsAws
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1

        Error: unable to delete HVN (test-perf-hvn-2-202301251411): context canceled waiting for delete HVN operation (43daf7c0-f1bf-4e0f-a4d4-43bd64e6b9b5) to complete


        Error: unable to delete HVN (test-perf-hvn-1-202301251411): context canceled waiting for delete HVN operation (0e0d2f89-c6d1-4e80-bab5-b594a387f9f6) to complete

--- FAIL: TestAccPerformanceReplication_ValidationsAws (12984.60s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-hcp/internal/provider	12985.886s
FAIL
```

**Review by codergs — COMMENTED** [review]
LGTM, thanks for the testing.
